### PR TITLE
Bonnie v2.3 restful calculation call

### PR DESCRIPTION
**Pull request by daco101**
This PR sets up the rest call and tests it (against the current calculation server which just echoes). The function will be modified once the cql calculation engine server is up and running. 
It also tweaks the test_helper so it can insert arrays (i.e. of value sets) into mongo.

Pull requests into Bonnie require the following. Submitter and reviewer should :white_check_mark: when done. For items that are not-applicable, note it's not-applicable ("N/A") and :white_check_mark:.

**Submitter:**
- [x] This pull request describes why these changes were made.
- [x] This PR is into the correct branch.
- [x] JIRA ticket for this PR: https://jira.mitre.org/browse/BONNIE-1577
- [x] JIRA ticket links to this PR
- [x] Code diff has been done and been reviewed (it **does not** contain: additional white space, not applicable code changes, debug statements, etc.)
- [x] If UI changes have been made, google WAVE plug-in has been executed to ensure no 508 issues were introduced.
- [x] Tests are included and test edge cases
- [x] Tests have been run locally and pass (remember to update Gemfile when applicable)
- [x] Test fixtures updated and documented as necessary ( see [internal wiki](https://gitlab.mitre.org/bonnie/internal-documentation/wikis/testing#test-fixtures) )
- [x] Code coverage has not gone down and all code touched or added is covered. 
     * In rare situations, this may not be possible or applicable to a PR. In those situations:
         1. Note why this could not be done or is not applicable here: 
         2. Add TODOs in the code noting that it requires a test
         3. Add a JIRA task to add the test and link it here: 

Branch | Back End Coverage | Front End Coverage
-- | -- | --
bonnie-v2.3-backend-calculation | 1903 / 2134 LOC (89.18%) covered| Statements   : 59.44% ( 7280/12247 )<br>Branches     : 43.43% ( 1979/4557 )<br>Functions    : 49.81% ( 1207/2423 )<br>Lines        : 57.31% ( 6536/11404 )
bonnie-v2.3-restful-calculation-call | 1911 / 2139 LOC (89.34%) covered | Statements   : 59.44% ( 7280/12247 ) <br> Branches     : 43.43% ( 1979/4557 ) <br> Functions    : 49.81% ( 1207/2423 ) <br> Lines        : 57.31% ( 6536/11404 )

- [x] Automated regression test(s) pass N/A

If JIRA tests were used to supplement or replace automated tests:
- [x] JIRA test links: N/A
- [x] Justification for using JIRA tests: N/A
- [x] JIRA tests have been added to sprint N/A

**Reviewer 1:**

Name: James
- [x] Code is maintainable and reusable, reuses existing code and infrastructure where appropriate, and accomplishes the task’s purpose
- [x] The tests appropriately test the new code, including edge cases
- [x] You have tried to break the code

If JIRA tests were used to supplement or replace automated tests:
- [x] JIRA tests have been run and pass
- [x] You agree with the justification for use of JIRA tests or have provided input on why you disagree

**Reviewer 2:**

Name: Chris
- [x] Code is maintainable and reusable, reuses existing code and infrastructure where appropriate, and accomplishes the task’s purpose
- [x] The tests appropriately test the new code, including edge cases
- [x] You have tried to break the code

If JIRA tests were used to supplement or replace automated tests:
- [x] JIRA tests have been run and pass
- [x] You agree with the justification for use of JIRA tests or have provided input on why you disagree

### CODE REVIEW

File: test/unit/backend_calc_test.rb [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+        prettyPrint: true
+      }
+      r = BonnieBackendCalculator.calculate(measure, patients, value_sets, options)
+
+      assert_equal measure.to_json, r['measure'].to_json
+      assert_equal value_sets.to_json, r['valueSets'].to_json
+      assert_equal options.to_json, r['options'].to_json
+
+      # patients are converted so we dont expect them to be echoed identically, just check a few fields
+      assert_equal patients.first['first'], r['patients'].first['givenNames'][0]
```
Comment (hossenlopp) [inline test/unit/backend_calc_test.rb]: This one still has ordering issues but for some reason checking for count of the arrays fix the ordering issue. Just add this before this line and it should work for people on mongo 3.4.4.
```
assert_equal patients.count, r['patients'].count
```

**Review by hossenlopp — CHANGES_REQUESTED** [review]
Just one last change to fix test ordering issues.

File: test/unit/backend_calc_test.rb (hunk 1 above)
Comment (hossenlopp) [reply test/unit/backend_calc_test.rb]: Nevermind. This doesn't work.

Comment (hossenlopp) [reply test/unit/backend_calc_test.rb]: Got a different suggestion. This grabs all the names and sorts them before checking the lists.
```
assert_equal (patients.collect { |p| p.first }).sort, (r['patients'].collect { |p| p['givenNames'][0] }).sort
assert_equal (patients.collect { |p| p.last }).sort, (r['patients'].collect { |p| p['familyName'] }).sort
```

**Review by hossenlopp — APPROVED** [review]
The error checking situations aren't tested, but that will be much easier to do once the service is running with calculation stuff intact to throw errors.

**Comment by hossenlopp** [comment]
Approving with travis failure due to nokogiri vulnerability that we will solve on master.

**Review by hossenlopp — CHANGES_REQUESTED** [review]
Just add a test to make sure timeouts to the service are handled gracefully.

**Review by hossenlopp — APPROVED** [review]

**Review by jbradl11 — APPROVED** [review]
Code looks good! Approving, knowing we will fix vulnerability in the base branch.

**Pull request merged**